SOSS/dms/soss_centroids.py

```diff
@@ -0,0 +1,377 @@
+
+    # Fit the y-positions with a polynomial and use the result as the true y-positions.
+    xtrace = np.arange(dimx)
+    mask = np.isfinite(ytrace)
+
+    # For padded arrays ignore padding for consistency with real data
+    if padding != 0:
+        mask = mask & (xtrace >= xos*padding) & (xtrace < (dimx - xos*padding))
+
+    # If no polynomial order was given return the raw measurements.
+    if poly_order is None:
+        param = []
+    else:
+        param = robust_polyfit(xtrace[mask], ytrace[mask], poly_order)
+        ytrace = np.polyval(param, xtrace)
+
+    # If verbose visualize the result.
+    if verbose is True:
+        _plot_centroid(image_masked, xtrace, ytrace)
+
+    return xtrace, ytrace, param
+
+
+def test_uncontam_centroids():
+    """"""
+
+    filename = os.path.join(PATH, 'SOSS_ref_2D_profile_SUBSTRIP256.fits.gz')
+
+    image, header = fits.getdata(filename, ext=2, header=True)
+    xtrace, ytrace, param = get_uncontam_centroids(image, header=header, verbose=True)
+
+    return
+
+
+def edge_trigger(image, halfwidth=5, yos=1, verbose=False):
+    """Detect the edges and center of the trace based on the minima and maxima of the derivate
+     of the columns, which is computed in a running window along the columns of the detector image
+
+     :param image: A 2D image of the detector.
+     :param halfwidth: the size of the window used when computing the derivatives.
+     :param yos: the oversampling factor of the image array along the y-direction.
+     :param verbose: If set True some diagnostic plots will be made.
+
+     :type image: array[float]
+     :type halfwidth: int
+     :type yos: int
+     :type verbose: bool
+
+     :returns: ytrace_max, ytrace_min, ytrace_comb - The upper edge, lower edge and center of the trace.
+     :rtype: Tuple(array[float], array[float], array[float])
+     """
+
+    dimy, dimx = image.shape
+    halfwidth = halfwidth * yos
+
+    # Create coordinate arrays.
+    xpix = np.arange(dimx)
+    ypix = np.arange(dimy)
+    _, ygrid = np.meshgrid(xpix, ypix)
+
+    # Compute windowed slopes over the columns.
+    slopevals = np.zeros_like(image)
+    for irow in range(halfwidth, dimy-halfwidth):
+
+        # Compute the window indices.
+        ymin = irow - halfwidth
+        ymax = irow + halfwidth + 1
+
+        # Get the x and y data to find the slope to.
+        datay = image[ymin:ymax, :]
+        mask = np.isfinite(datay)
+        datax = np.where(mask, ygrid[ymin:ymax, :], np.nan)  # Need to set values NaN in y to NaN in x.
+
+        # Compute the slope.
+        with warnings.catch_warnings():
+            warnings.simplefilter("ignore")
+            xmean = np.nanmean(datax, axis=0, keepdims=True)
+            ymean = np.nanmean(datay, axis=0, keepdims=True)
+
+        with np.errstate(invalid='ignore'):
+            slope = np.nansum((datax - xmean) * (datay - ymean), axis=0) / np.nansum((datax - xmean) ** 2, axis=0)
```

Comment:
This line is too long for my comparison! I'm going to keep bugging you about 80 character limit, so I can read this side-by-side

Reply:
And yet my PyCharm defaults to 120 characters. Can we please forget the punchcard era? 
Anyway split into numerator and denominator.